File: cgo_simple/mybuffer/cgo/my_buffer.go
package cgo

import "unsafe"

type MyBuffer struct {
	cptr CgoMyBufferHandle
}

func NewMyBuffer(size int) *MyBuffer {
	return &MyBuffer{
		cptr: cgo_NewMyBuffer(size),
	}
}

func (this *MyBuffer) Delete() {
	cgo_DeleteMyBuffer(this.cptr)
}

func (this *MyBuffer) Data() []byte {
	data := cgo_MyBuffer_Data(this.cptr)
	len := cgo_MyBuffer_Size(this.cptr)
	return ((*[1 << 31]byte)(unsafe.Pointer(data)))[0:int(len):int(len)]
}

File: cgo_simple/mybuffer/capi/MyBuffer.cpp
#include "MyBuffer.h"

MyBuffer::MyBuffer(int iSize){
    this->m_data = new std::string(iSize,'\0');
}

MyBuffer::~MyBuffer(){

}

char* MyBuffer::Data(){
    return (char*)this->m_data->data();
}

int MyBuffer::Len() const {
    return this->m_data->length();
}